### PR TITLE
feat: allow to run multiple celery queues and workers

### PR DESCRIPTION
**Pull request by Ian2012**
### Description

This PR allows the deployment of multiple celery queues and workers via settings. It also overrides the default lms-worker deployment to prevent it to consume events from all queues.

Deployed on `musashi/tunis`

### Future work

- Having a special queue/deployment for Aspects
- Support extra arguments for the deployments
- Support the `gevent` pool (install `gevent` on openedx via this plugin)
- Add a tutor filter for the workers

### CODE REVIEW

File: tutorcelery/patches/k8s-deployments [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        - name: {{service}}-worker-{{variant}}
+          image: {{ DOCKER_IMAGE_OPENEDX }}
+          args:
+            - "celery"
+            - "--app={{service}}.celery"
+            - "worker"
+            - "--loglevel=info"
+            - "--hostname=edx.{{service}}.core.{{variant}}"
+            - "--max-tasks-per-child=100"
+            - "--queue=edx.{{service}}.core.{{variant}}"
```
Comment (jfavellar90) [inline tutorcelery/patches/k8s-deployments]:
```suggestion
            - "--queues=edx.{{service}}.core.{{variant}}"
```
is it `queue` or `queues` (or do both work?)

File: tutorcelery/patches/k8s-deployments [hunk 2]
```diff
@@ -0,0 +1,55 @@
+{% for service, variants in CELERY_WORKER_VARIANTS.items() %}
+
```
Comment (jfavellar90) [inline tutorcelery/patches/k8s-deployments]: Is this extra empty line required?

File: tutorcelery/patches/k8s-deployments [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            allowPrivilegeEscalation: false
+      volumes:
+        - name: settings-{{service}}
+          configMap:
+            name: openedx-settings-{{service}}
+        - name: config
+          configMap:
+            name: openedx-config
+{% endfor %}
+
```
Comment (jfavellar90) [inline tutorcelery/patches/k8s-deployments]: Is this extra empty line required?

File: tutorcelery/patches/k8s-override [hunk 4]
```diff
@@ -0,0 +1,22 @@
+{% for service in ["lms", "cms"] %}
+
```
Comment (jfavellar90) [inline tutorcelery/patches/k8s-override]: Is this extra empty line required?

File: tutorcelery/patches/k8s-deployments [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        runAsGroup: 1000
+      containers:
+        - name: {{service}}-worker-{{variant}}
+          image: {{ DOCKER_IMAGE_OPENEDX }}
+          args:
+            - "celery"
+            - "--app={{service}}.celery"
+            - "worker"
+            - "--loglevel=info"
+            - "--hostname=edx.{{service}}.core.{{variant}}"
```
Comment (jfavellar90) [inline tutorcelery/patches/k8s-deployments]:
```suggestion
            - "--hostname=edx.{{service}}.core.{{variant}}.%%h"
```
Should it be like the [Tutor hostname](https://github.com/overhangio/tutor/blob/master/tutor/templates/k8s/deployments.yml#L144C110-L144C113)?

File: tutorcelery/patches/k8s-override [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    spec:
+      containers:
+        - name: {{service}}-worker
+          image: {{ DOCKER_IMAGE_OPENEDX }}
+          args:
+            - "celery"
+            - "--app={{service}}.celery"
+            - "worker"
+            - "--loglevel=info"
+            - "--hostname=edx.{{service}}.core.default"
```
Comment (jfavellar90) [inline tutorcelery/patches/k8s-override]:
```suggestion
            - "--hostname=edx.{{service}}.core.default.%%h"
```
Should it be like the [Tutor hostname](https://github.com/overhangio/tutor/blob/master/tutor/templates/k8s/deployments.yml#L144C110-L144C113)?

File: tutorcelery/patches/k8s-override [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+            - "--max-tasks-per-child=100"
+            - "--queues=edx.{{service}}.core.default"
```
Comment (jfavellar90) [inline tutorcelery/patches/k8s-override:19]: `queue` or `queues`?

Comment (jfavellar90) [reply tutorcelery/patches/k8s-override:19]: Apparently is [queues](https://docs.celeryq.dev/en/stable/reference/cli.html#cmdoption-celery-worker-Q)

File: tutorcelery/patches/openedx-common-settings [hunk 8]
```diff
@@ -0,0 +1 @@
+EXPLICIT_QUEUES.update({{CELERY_EXPLICIT_QUEUES}})
```
Comment (jfavellar90) [inline tutorcelery/patches/openedx-common-settings]: Rendering the `CELERY_EXPLICIT_QUEUES` variable this way can lead to errors. We need to render this in a format Python can understand. One option could be:

```suggestion
EXPLICIT_QUEUES.update(json.loads('{{CELERY_EXPLICIT_QUEUES | tojson}}'))
```
but importing `json` is annoying. Other option could be the iteration over the object to assign every key to the corresponding value

Comment (jfavellar90) [reply tutorcelery/patches/openedx-common-settings]: The other option is to remove this feature so `EXPLICIT_QUEUES` is not modified

Comment (Ian2012) [reply tutorcelery/patches/openedx-common-settings]: When using python files, this gets rendered correctly, we can tweak the patch to render it correctly

Comment (jfavellar90) [reply tutorcelery/patches/openedx-common-settings]: From what we saw, it renders OK

File: tutorcelery/patches/openedx-lms-production-settings [hunk 9]
```diff
@@ -0,0 +1,4 @@
+try:
+    EXPLICIT_QUEUES.update({{CELERY_EXPLICIT_QUEUES}})
+except AttributeError:
```
Comment (jfavellar90) [inline tutorcelery/patches/openedx-lms-production-settings]: Should we catch `NameError` if the variable is not defined?

**Comment by jfavellar90** [comment]
@Ian2012 I remain attentive to the results of the tests adding LMS tasks to CMS configuration (via EXPLICIT_QUEUES) and vice-versa

**Comment by Ian2012** [comment]
@jfavellar90 Mixing the settings for the tasks routes can cause different issues, I've decided to create two separate variables to control LMS and CMS workers tasks routing.

File: README.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    - high
+    - high_mem
+    - lms_custom_queue
+  cms:
+    - high
+    - low
+    - cms_custom_queue
+```
+
+This plugin also provides a setting to directly route LMS/CMS tasks to an specific queue:
```
Comment (jfavellar90) [inline README.md]: I think it's important to mention that this will extend/override the variable `EXPLICIT_QUEUES` on both LMS and CMS

**Review by jfavellar90 — APPROVED** [review]

**Comment by jfavellar90** [comment]
I already provided my approval @Ian2012, however, please fix the minor detail I mentioned in my last comment and proceed with the merge.

**Pull request merged**